Fix DHCPv4 tests

rveerama1 opened the pull request:
net_pkt_alloc_with_buffer() takes IP header and protocol header length while calculating total length internally. Need not specify explicitly. Also mutex was not properly initialized.

rveerama1 committed:
net: tests: Fix DHCPv4 tests

net_pkt_alloc_with_buffer() takes IP header and protocol header
length while calculating total length internally. Need not
specify explicitly. Also mutex was not properly initialized.

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

pfalcon commented on tests/net/dhcpv4/src/main.c:
```diff
@@ -156,14 +156,12 @@ struct dhcp_msg {
+/* As per subsys/net/ip/dhcpv4.c
+ * "RFC2131 4.1.1 requires we wait a random period between 1 and 10
+ * seconds before sending the initial discover". So test will wait
+ * max 11 seconds.
```
What? Test waits up to 11 seconds? That's too long. If even small percentage of tests will do tricks like that, the testsuite will take enormous time to run (and we have regulaa CI timeouts already). Please limit that to 2s, max 3s. If that requires delay in the stack code to be configurable, well, do that, and override it in test config.

rveerama1 replied:
https://github.com/rveerama1/zephyr/blob/master/subsys/net/ip/dhcpv4.c#L1056

rveerama1 replied:
Otherwise, I have to introduce new config option for DHCPv4 testing and make it 0 to 2s max.

pfalcon replied:
> Otherwise, I have to introduce new config option for DHCPv4 testing and make it 0 to 2s max

Yes, that's what I suggest, make DHCPV4_INITIAL_DELAY_MAX into CONFIG_DHCPV4_INITIAL_DELAY_MAX. Because today I ran some POSIX test, and it also ran noticeable number of seconds. If we let those expand uncontrollably, it will be havoc. Also ping @nashif, @galak for suggestions.

rveerama1 replied:
Proposed some changes, please take a look.

pfalcon replied:
> Proposed some changes, please take a look.

Looks good. I'd personally just make that a normal option to cut on the number of if's, after all, if someone will want to break standard, they'll do that with or without option.

But you introduced the option, but not set in the test config, so test fails.

rveerama1 replied:
depends on NET_DHCPV4
	depends on ZTEST
	default 2
Isn't it enabled by default if NET_DHCPV4 and ZTEST are enabled.
Without specifying in test prj.conf I can see it in generated .config file.
CI fails, I am investigating it.

rveerama1 replied:
> Looks good. I'd personally just make that a normal option to cut on the number of if's, after all, if someone will want to break standard, they'll do that with or without option.

Ok, I will try that.

rveerama1 committed:
net: dhcpv4: Introduce config option for max delay

As per RFC2131 4.1.1 requires we wait a random period
between 1 and 10 seconds before sending the initial
discover. But tests can not wait that longer. So this
option helps test to configure the value to minimum.

Signed-off-by: Ravi kumar Veeramally <ravikumar.veeramally@linux.intel.com>

rveerama1 commented:
@pfalcon take a look of new version.

jukkar reviewed (approved):
LGTM
